question-servers/freeform.js

```diff
@@ -1253,535 +1217,559 @@ module.exports = {
     const courseIssues = [];
     let panelCount = 0,
       cacheHitCount = 0;
-    module.exports.getContext(question, course, (err, context) => {
-      if (ERR(err, callback)) return;
-
-      workers.getPythonCaller((err, pc) => {
-        if (ERR(err, callback)) return;
-        async.series(
-          [
-            // FIXME: support 'header'
-            async () => {
-              if (!renderSelection.question) return;
-
-              const {
-                courseIssues: newCourseIssues,
-                html,
-                renderedElementNames,
-                cacheHit,
-              } = await module.exports.renderPanelInstrumented(
-                'question',
-                pc,
-                submission,
-                variant,
-                question,
-                course,
-                locals,
-                context
-              );
+    const context = await module.exports.getContext(question, course);
 
-              courseIssues.push(...newCourseIssues);
-              htmls.questionHtml = html;
-              panelCount++;
-              if (cacheHit) cacheHitCount++;
-              allRenderedElementNames = _.union(allRenderedElementNames, renderedElementNames);
-            },
-            async () => {
-              if (!renderSelection.submissions) return;
-
-              htmls.submissionHtmls = await async.mapSeries(submissions, async (submission) => {
-                const {
-                  courseIssues: newCourseIssues,
-                  html,
-                  renderedElementNames,
-                  cacheHit,
-                } = await module.exports.renderPanelInstrumented(
-                  'submission',
-                  pc,
-                  submission,
-                  variant,
-                  question,
-                  course,
-                  locals,
-                  context
-                );
+    return workers.withPythonCaller(async (pc) => {
+      await async.series([
```

Comment:
Do we actually need this `series()` here? I can't see what it's doing.

Reply:
Nope! I'll remove it.

Reply:
This was leftover from some refactoring where I was still mixing async and callback-based code. I removed it in 1b39372.

Reply:
Actually, I take it back, it was used - we use this to skip rendering with things like the following:

```ts
if (!renderSelection.question) return;
```

We could refactor that to use `if (renderSelection.question) { ... }` instead, but I'm inclined to keep this as is for now.

Reply:
Sure, I'm on board with that.